### ISSUE TITLE
Slight race with cache invalidation of Raft truncated state

### ISSUE THREAD

**Issue by kkaneda** [cockroachdb/cockroach#2815]
`TestReplicateRogueRemovedNode()` gets the following panic if we insert `time.Sleep(1*time.Second)` just before `setCachedTruncatedState()` is called in `applyRaftCommand()`.

```
panic: requested entry at index is unavailable

goroutine 6376 [running]:
github.com/coreos/etcd/raft.(*raftLog).term(0xc8200e2bd0, 0x16, 0x42af028, 0x0, 0x0)
        /Users/kaneda/Development/go/src/github.com/coreos/etcd/raft/log.go:231 +0x1a9
github.com/coreos/etcd/raft.(*raftLog).matchTerm(0xc8200e2bd0, 0x16, 0x6, 0xc820123cc0)
  ...
```

I think there is a slight chance that Raft gorountine accesses the cached truncated state and attempts to access deleted logs before the cache is invalidated.

Example scenario:
1) Replica executes `TruncateLog`
2) The batch is committed and the log is truncated.
3) Before the in-memory `truncatedState` is updated, Raft Multinode reads the cached state (e.g., calls `FirstIndex()`) and gets the stale index/term.

**Comment by bdarnell** [comment]
I can believe that there is a race in `setCachedTruncatedState`, but `TestReplicateRogueRemovedNode` never calls `TruncateLog` (I think only one of our tests does). The race in that test must be somewhere else. 

Where exactly did you add the sleep? Inside the `ba.GetArg(roachpb.TruncateLog)` if statement, or before?

**Comment by kkaneda** [comment]
Hmm, you're right. I got a similar panic while modifying the other PR. 

It seems tests under `storage` look flaky now. (`go test` under `storage` doesn't pass for me with the latest master.) 

Yes, inside the `if` statement:

```
		if _, ok := ba.GetArg(roachpb.TruncateLog); ok {
			time.Sleep(1*time.Second)
			r.setCachedTruncatedState(nil)
		}
```

**Comment by kkaneda** [comment]
Ah, please ignore. That might be related to my local env... I was using master of `coreos/etcd`.

**Comment by kkaneda** [comment]
Looked into more. This seems like a separate bug.

A panic I got was coming from `raftLog.appliedTo()`, which was called from `raft.newRaft()`. It seems `c.Applied` was set from a value of `Replica.appliedIndex` while the `log.applied` was set from a value obtained from the first index (= `Index` of the raft truncated state). 

1) `state.createGroup()` is called. Call `s.StateMachine.AppliedIndex()` and pass the value to `raftCfg.Applied`. The value is 21.
2) `raft.newRaft()` calls `newLog()`, which calls `storage.FirstIndex()`. 1 is returned since no raft truncated state was written and the replica hadn't yet been initialized. `log.applied` is set to 0.
3) `raftLog.appliedTo` fails since `log.applied` is smaller than `raftCfg.Applied`.

I don't know the right fix, but I guess we can do something like:

- change `raft.newLog` to set `log.applied` from a given `Config`'s `Applied`

or 

- change `state.createGroup()` to set `raft.Config.Applied` from the first index

or 

- set the index of the raft truncated state to the value of `Replica.appliedIndex` when creating a new replica.

**Comment by bdarnell** [comment]
`Store.AppliedIndex` and `Replica.FirstIndex` (and `Replica.LastIndex`) should be consistent with each other: `AppliedIndex` should return 0 for uninitialized ranges. But how did we get an applied index of 21 without initializing the range? We should investigate that before adding an `IsInitialized` check in `AppliedIndex`.

It could be a race in `ApplySnapshot`: we set the `appliedIndex` field before updating the descriptor (and we're using a bunch of separate atomic updates instead of a lock).

**Comment by kkaneda** [comment]
Ah, I see. So, it seems the root cause was same as #2802 . The failure stopped happening with the recent master. I tweaked code a bit more if I can still reproduce some race, but couldn't.

**Comment by bdarnell** [comment]
#2802 was a dupe of #1878, which is now fixed, so maybe that's why it stopped happening. But we never saw exactly these symptoms before, so maybe it's more like #2837, which is still open and apparently related to multi-cpu configurations.

**Comment by kkaneda** [comment]
Ah, interesting. I was able to reproduce this with reasonable probability if I use an old branch. Still looking into, but here is a log snippet with additional log messages. It indicates that Store ID 2 set the Raft hard state, but it failed to load it later when creating a new Raft instance.

```
I1014 23:10:42.865194 91428 storage/replica_raftstorage.go:495  [2] setting the hard state: range=1, commit=21
I1014 23:10:42.865406 91428 storage/replica_raftstorage.go:208  [2] setting applied index to 21
I1014 23:10:42.865954 91428 storage/replica_raftstorage.go:44  [2] initial state: range=1, found=true
E1014 23:10:42.866023 91428 storage/client_test.go:492  error reading "a" from engine 1: value type is not INT: UNKNOWN
I1014 23:10:42.866061 91428 storage/replica_raftstorage.go:44  [2] initial state: range=1, found=false
W1014 23:10:42.866094 91428 raft/raft.go:208  [group 1] no hard state found
E1014 23:10:42.866115 91428 raft/log.go:195  [group 1] applied(21) is out of range [prevApplied(0), committed(0)]
panic: [group 1] applied(21) is out of range [prevApplied(0), committed(0)]
```

It looks strange that somehow we lost the hard state. There was no simultaneous snashot.

**Comment by kkaneda** [comment]
It seems what happened was that we attempted to create a new Raft group while a replica was being destroyed by the GC queue. 

1) Raft loads the applied index to start a new Raft Group (via `Store.AppliedIndex`).
2) `Replica.Destroy()` is called. The Raft hard state is cleaned up.
3) `raft.newRaft()` is called. `c.Storage.InitialState()` returns an empty state, so `r.loadState()` is not called.
4) `raftlog.appliedTo` fails.

**Comment by bdarnell** [comment]
Yep, that's definitely a race. And there's even a `TODO(bdarnell)` for it: https://github.com/cockroachdb/cockroach/blob/9515804b66f4cfc125675aa45909559565ffbf40/storage/replica_gc_queue.go#L127

I have a fix in the works, although I still see the "applied out of range" panic on rare occasions (I increased the iteration count in `TestRaftRemoveRace` and the error occurs about once every 10k iterations).

**Comment by bdarnell** [comment]
The most common issues here should have been solved in #2868; I've filed a couple more issues to track the remaining problems: #2877 and #2878.
